numba/targets/arraymath.py

```diff
@@ -999,3 +1019,186 @@ def dot_3(context, builder, sig, args):
+
+    This function compute the LU decomposition of a matrix.
+
+    """
+    fnty = ir.FunctionType(ir.IntType(32),
+                           [ll_char,                       # kind
+                            intp_t, intp_t,                # m, n
+                            ll_void_p, intp_t,             # a, lda
+                            ll_intc_p, ll_intc_p           # ipiv, info
+                           ])
+
+    fn = builder.module.get_or_insert_function(fnty, name="numba_xxgetrf")
+
+    kind = get_blas_kind(a_type.dtype)
+    kind_val = ir.Constant(ll_char, ord(kind))
+
+    if a_type.layout == 'F':
+        m, n = a_shapes
+        lda = a_shapes[0]
+    else:
+        n, m = a_shapes
+        lda = a_shapes[1]
+
+    res = builder.call(fn, (kind_val, m, n,
+                            builder.bitcast(a_data, ll_void_p), lda,
+                            ipiv, info
+                            ))
+    check_lapack_return(context, builder, res)
+
+
+def call_xxgetri(context, builder, a_type, a_shapes, a_data, ipiv, work,
+                 lwork, info):
+    """
+    Call the LAPACK gettri function for the given arguemnt.
+
+    This function compute the invers of a matrix given its LU decomposition.
+
+    """
+    fnty = ir.FunctionType(ir.IntType(32),
+                           [ll_char,                       # kind
+                            intp_t, ll_void_p, intp_t,     # n, a, lda
+                            ll_intc_p, ll_void_p,          # ipiv, work
+                            ll_intc_p, ll_intc_p           # lwork, info
+                           ])
+    fn = builder.module.get_or_insert_function(fnty, name="numba_xxgetri")
+
+    kind = get_blas_kind(a_type.dtype)
+    kind_val = ir.Constant(ll_char, ord(kind))
+
+    n = lda = a_shapes[0]
+
+    res = builder.call(fn, (kind_val, n,
+                            builder.bitcast(a_data, ll_void_p), lda,
+                            ipiv, builder.bitcast(work, ll_void_p),
+                            lwork, info
+                            ))
+    check_lapack_return(context, builder, res)
+
+
+def mat_inv(context, builder, sig, args):
+    """
+    Invert a matrix by first computing its LU decomposition.
+    """
+    xty = sig.args[0]
+    dtype = xty.dtype
+
+    x = make_array(xty)(context, builder, args[0])
+    x_shapes = cgutils.unpack_tuple(builder, x.shape)
+    m, n = x_shapes
+    check_c_int(context, builder, m)
+    check_c_int(context, builder, n)
+
+    # Allocate the return array (Numpy never works in place contrary to
+    # Scipy for which one can specify to whether or not to overwrite the
+    # input.
+    def create_out(a):
+        m, n= a.shape
+        if m != n:
+            raise ValueError("np.linalg.inv can only work on square "
+                             "arrays.")
+        return a.copy()
+
+    out = context.compile_internal(builder, create_out,
+                                   signature(xty, *sig.args), args)
```

Comment:
This function does not return an array of type `xty`, but of type `sig.return_type` (the layout may differ, which is probably the cause of the error with 'F' arrays).